extender.go

```diff
@@ -113,38 +121,272 @@ func (f *ExtenderFactory) verifyAPIs(groupExt []buildpack.GroupElement, logger l
+		return fmt.Errorf("getting image layers: %w", err)
+	}
+	var (
+		currentHash  v1.Hash
+		needsCopying bool
+	)
+	if origTopLayerHash == "" { // if the original base image had no layers, copy all the layers
+		needsCopying = true
+	}
+	group, _ := errgroup.WithContext(context.TODO())
+	for _, currentLayer := range layers {
+		currentHash, err = currentLayer.Digest()
 		if err != nil {
-			return err
+			return fmt.Errorf("getting layer hash: %w", err)
 		}
-		if buildDockerfile != nil {
-			e.Logger.Debugf("Found build Dockerfile for extension '%s'", ext.ID)
-			dockerfiles = append(dockerfiles, *buildDockerfile)
+		switch {
+		case needsCopying:
+			currentLayer := currentLayer // allow use in closure
+			group.Go(func() error {
+				return copyLayer(currentLayer, toPath)
+			})
+		case currentHash.String() == origTopLayerHash:
+			needsCopying = true
+			continue
+		default:
+			continue
 		}
 	}
-	options := extend.Options{
-		CacheTTL:    e.CacheTTL,
-		IgnorePaths: []string{e.AppDir, e.LayersDir, e.PlatformDir},
+	return group.Wait()
+}
+
+func copyLayer(layer v1.Layer, toSparseImage string) error {
+	digest, err := layer.Digest()
+	if err != nil {
+		return err
+	}
+	f, err := os.Create(filepath.Join(toSparseImage, "blobs", digest.Algorithm, digest.Hex))
+	if err != nil {
+		return err
 	}
-	digest, err := name.NewDigest(e.ImageRef)
+	defer f.Close()
+	rc, err := layer.Compressed() // FIXME: if exporting to a daemon, this should be uncompressed
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+	_, err = io.Copy(f, rc)
+	return err
+}
+
+const (
+	argBuildID = "build_id"
+	argUserID  = "user_id"
+	argGroupID = "group_id"
+)
+
+func (e *Extender) extend(kind string, baseImage v1.Image, logger log.Logger) (v1.Image, error) {
+	logger.Debugf("Extending base image for %s: %s", kind, e.ImageRef)
+	dockerfiles, err := e.dockerfilesFor(kind, logger)
 	if err != nil {
-		return fmt.Errorf("failed to get digest for reference %s: %w", e.ImageRef, err)
+		return nil, fmt.Errorf("getting %s dockerfiles: %w", kind, err)
+	}
+
+	var (
+		configFile *v1.ConfigFile
+		rebasable  = true // for now, we don't require the initial base image to have io.buildpacks.rebasable=true
+	)
+	configFile, err = baseImage.ConfigFile()
+	if err != nil || configFile == nil {
+		return nil, fmt.Errorf("getting image config: %w", err)
 	}
-	return e.DockerfileApplier.Apply(e.AppDir, digest.DigestStr(), dockerfiles, options)
+	buildOptions := e.extendOptions()
+	for _, dockerfile := range dockerfiles {
+		userID, groupID := userFrom(*configFile)
+		dockerfile.Args = append([]extend.Arg{
+			{Name: argBuildID, Value: uuid.New().String()},
+			{Name: argUserID, Value: userID},
+			{Name: argGroupID, Value: groupID},
+		}, dockerfile.Args...)
+
+		// apply Dockerfile
+
+		if baseImage, err = e.DockerfileApplier.Apply(
+			dockerfile,
+			baseImage,
+			buildOptions,
+			logger,
+		); err != nil {
+			return nil, fmt.Errorf("applying dockerfile to image: %w", err)
```

Comment:
if/when we return partway through this loop what does that mean for the user's docker images?  are they going to be partially extended and in an inconsistent or unusable state?  is there a way to ... "roll back the transaction"?   or maybe indicate to the user "our condolences for the loss of your docker image. you have a backup, right?"   Or is the UX cool, here?

Reply:
Great question! It should be okay - we're using the filesystem as our export target. Specifically, we're writing to the kaniko dir as a "scratch" workspace and only copying the image layers to `<extended>/run/<image>` after all Dockerfiles have been applied.